docs/BACKEND.md

```diff
@@ -114,4 +114,65 @@ matter what language or framework crates.io was implemented in.
 
 ## Tests
 
+### Integration tests
+
+A suite of integration tests that require a full crates.io app to be
+instantiated and available live in `src/tests`. These are a mixture of tests
+that exercise routes and controllers like normal API consumers, and other tests
+that require a full blown application and database to be available.
+
+Tests that interact with HTTP services — for example, S3 to upload packages or
+index entries — do so by routing requests through a HTTP proxy configured within
+`TestApp`. This proxy implements similar behaviour to the [`vcr` Ruby
+gem](https://github.com/vcr/vcr): by default, requests are not forwarded to the
+actual upstream service, but instead the request is checked against a
+"cassette", which is a JSON file in `src/tests/http-data` that contains the
+expected request and response. If the request matches, then the saved response
+is returned. If the request doesn't match, then an error is returned and the
+test fails.
+
+#### Updating test cassettes
+
+When updating integration tests that make HTTP requests, you may need to update
+the test cassette associated with that test. This is controlled by the `RECORD`
+environment variable: when set to `yes`, the request will actually be forwarded
+to the upstream service, and the cassette will be updated with the new response.
+
+In addition to setting `RECORD`, you will also need to set a number of other
+environment variables to handle any requests to S3. These are in `.env.sample`,
+but are also reproduced here for convenience:
+
+* `TEST_S3_BUCKET`: the S3 bucket used for package uploads.
+* `TEST_S3_REGION`: the S3 region used to package uploads. This may also be an
+  absolute URL (such as `http://127.0.0.1:19000`), in which that will be used as
+  the endpoint.
```

Comment:
does it make sense for this to be configurable? the URL ends up in the JSON files, so using something different would cause the tests to fail (see below) 🤔

Reply:
I think so: what it does do is allow a test run against real S3 with `RECORD=yes` — while the result of that isn't something we'd want to commit, it means we can do a test run against the real service and be comfortable that we still get real responses back.